TestRunners/TestRunner.py

```diff
@@ -0,0 +1,181 @@
+        """ Handles user input of test folder name and flags as described by 
+        system_args and runs tests accordingly. The method does the following:
+        1) Checks if test folder input refers to an existing directory and returns the flag manager's to_string if it does not
+        2) Checks if input is -h or --help
+        3) Applies actions to test files as described by flags 
+        4) Calls before_tests_run
+        5) Starts iterating over c files in test directory
+        6) Calls on_tests_run 
+        7) Checks if tests has failed and skips to next test if it has
+        8) Calls on_test_will_run if test will run
+
+        The exit status is returned, with 0 indicating success and 1 indicating failure
+
+        Args:
+            system_args (list[str])
+
+        Returns:
+            int: exit staus (0 = success; 1 = failure)
+        """
+        if os.path.split(os.getcwd())[1] != "build":
+            print(f"ERROR: not in build dir; cwd is {os.getcwd()}")
+            sys.exit(1)
+
+        if len(system_args) == 1:
+            print(self._flags_manager.to_string())
+            return 1
+
+        # for sys.argv, 0th = ../run_test.py, 1st = folder name, 2 and beyond = user args
+        test_folder_name = system_args[1]
+
+        if (test_folder_name == "-h" or test_folder_name == "--help"):
+            print(self._flags_manager.to_string())
+            return 0
+
+        try:
+            c_files = get_all_c_files(f"../test/{test_folder_name}")
+        except Exception as e:
+            print(e)
+            print(self._flags_manager.to_string())
+            return 1
+
+        for c_file in c_files:
+            self._add_c_file(c_file)
+
+        sys_arg_iterator = 2
+        while sys_arg_iterator < len(sys.argv):
+            arg = sys.argv[sys_arg_iterator]
+
+            if (self._flags_manager.flag_exists(arg)):
+                flag = self._flags_manager.get_flag(arg)
+
+                action_arg_count = flag.get_action().__code__.co_argcount
+
+                if (action_arg_count == 1):
+                    flag.get_action()(self._test_files_manager)
+                elif (action_arg_count == 2):
+                    try:
+                        next_arg = sys.argv[sys_arg_iterator + 1]
+                    except IndexError:
+                        raise FileNotFoundError(
+                            f"c file not specified for flag {arg}")
+
+                    flag.get_action()(self._test_files_manager, next_arg)
+                    sys_arg_iterator += 2
+                    continue
+
+            else:
+                raise ValueError(
+                    f"Flag {arg} not found. {self._flags_manager.to_string()}")
+
+            sys_arg_iterator += 1
+
+        self._before_tests_run()
+
+        for c_file in c_files:
+            f = self._test_files_manager.get_file(
+                c_file[find_nth(c_file, "/", 3) + 1:])
```

Comment:
Please add a comment explaining the constant indices used in this string operation.